### PR TITLE
Always store the image digest when pulling an image

### PR DESCRIPTION
**Pull request by dnephin**
cc @aanand @icecrime @aaronlehmann 

Tested manually and it appears to be working the way I expect, but maybe there are some edge cases I'm not considering.

### CODE REVIEW

File: distribution/pull_v2.go [hunk 1]
```diff
@@ -403,10 +403,19 @@ func (p *v2Puller) pullV2Tag(ctx context.Context, ref reference.Named) (tagUpdat
...
-	} else if err = p.config.ReferenceStore.AddTag(ref, imageID, true); err != nil {
-		return false, err
-	}
+	} else {
+		if err = p.config.ReferenceStore.AddTag(ref, imageID, true); err != nil {
+			return false, err
+		}
 
+		// Best effort store the digest, this could fail if the digest was
+		// already saved from a different tag.
```
Comment (dnephin) [inline distribution/pull_v2.go]: This is mostly just an assumption based on reading the code. Haven't had a chance to test it with the error. Can anyone confirm?

File: distribution/pull_v2.go [hunk 2]
```diff
@@ -403,10 +403,19 @@ func (p *v2Puller) pullV2Tag(ctx context.Context, ref reference.Named) (tagUpdat
...
+		digest, err := reference.WithDigest(ref, manifestDigest)
+		if err != nil {
+			return true, nil
```
Comment (LK4D4) [inline distribution/pull_v2.go]: looks like a place for `logrus.Errorf`

**Comment by aaronlehmann** [comment]
The problem with storing a digest for each image is that the reference store actually stores a separate reference for the digest. Effectively the image is tagged twice, once with a descriptive tag and once with a digest. We have some hacks in the UX to hide this slightly (i.e. deleting the tag when the only remaining reference is a digest will also delete the digest), but it's still rather kludgy and it's a source of confusion. We had hoped to find a solution to this before unconditionally storing digests in the reference store.

The best way forward is probably to add support for the `name:tag@digest` form so a single reference can embody both a tag and a digest. This would have a backward compatibility impact on the way references are stored, and possibly on the client protocol.

cc @dmcgowan @endophage @stevvooe

File: distribution/pull_v2.go (hunk 2 above)
Comment (dnephin) [inline distribution/pull_v2.go]: true, this should log an error

**Comment by dnephin** [comment]
> Effectively the image is tagged twice, once with a descriptive tag and once with a digest.

Could you elaborate on why that is a problem?

The problem I'm trying to solve is that we need the `RepoDigests` for pulled images.

Don't we end up in the same situation if someone does this:

```
$ docker pull alpine:3.3
# copy digest from output
$ docker pull alpine@sha256:45f2f6f12c7b2685032abb537aa89916b895e13f4cbe52cf22773e460577c69d
```

That's what we're forced to do right now, but it feels like it shouldn't be necessary to do the second pull.

**Comment by aaronlehmann** [comment]
> Could you elaborate on why that is a problem?

Here are some examples of how it can be confusing:

```
$ DOCKER_CONTENT_TRUST=1 docker pull nginx
$ docker images | grep nginx
nginx                    latest              0d409d33b27e        12 days ago         182.8 MB
nginx                    <none>              0d409d33b27e        12 days ago         182.8 MB
# Wait, why do I have two nginx entries now??
$ docker tag nginx my-nginx
$ docker rmi nginx
Untagged: nginx:latest
$ docker images | grep nginx
my-nginx                 latest              0d409d33b27e        12 days ago         182.8 MB
nginx                    <none>              0d409d33b27e        12 days ago         182.8 MB
# Super weird, I rmi'd nginx but I still see an nginx image. How do I delete that one??
```

I believe there are some GitHub issues related to content trust that raise these issues with image management, though I don't have the issue numbers handy. I don't have any strong feelings about this, but I understood the consensus was that we'd have to find a way to avoid duplicate image entries for every pull before we could enable content trust by default.

also cc @icecrime

**Comment by aaronlehmann** [comment]
> is it necessary to have compatibility implications? Would it not be possible to support name:tag@digest as an additional format or would that introduce too much unwanted complexity?

What I mean is that introducing this format in the references file will make that file unusable when downgrading to an older version. Perhaps this is not considered a problem.

**Comment by endophage** [comment]
Ah, that's forwards compatibility. I'm not aware of many applications that promise an old client will understand new formats, but not my decision to make here.

**Comment by stevvooe** [comment]
LGTM

**Comment by dmcgowan** [comment]
I agree with @aaronlehmann, we have to fix the UI first.  I am willing to help out with this to get this feature added.

We should consider these updates...
- `docker rmi` to also remove the digest if there no other tags with the same image name. So in the example Aaron showed the digest would be gone after a tag + rmi.
- `docker tag` to additionally create the digest reference if the source image was by digest.
- `docker images` does not show digest only references if there is a tag reference.

**Comment by dnephin** [comment]
@dmcgowan sgtm. I started looking into the `docker images` fix, but I couldn't find an easy solution. I guess we have to store all the images in a map by image id and then use it to filter out the duplicates?

**Comment by dmcgowan** [comment]
@dnephin: I think what the API is returning is already reasonable as well as group by image id. The output should just be displayed differently here https://github.com/docker/docker/blob/master/api/client/formatter/formatter.go#L212.
It should not treat each tag AND digest as always a new output line.

**Comment by dmcgowan** [comment]
Status update
- `docker images`  update - merged #23549
- `docker rmi` update - under review #23649 
- `docker tag` update - in progress but I don't think this one needs to block this PR since it is an unrelated corner case

Once the `rmi` change gets merged and the tests are passing this should be ready to go. I anticipate there may need to be some integration cli tests still updated, will know once the changes have been integrated.

**Comment by dnephin** [comment]
Build is green!

File: distribution/pull_v2.go [hunk 3]
```diff
@@ -403,10 +403,20 @@ func (p *v2Puller) pullV2Tag(ctx context.Context, ref reference.Named) (tagUpdat
...
+	} else {
+		if err = p.config.ReferenceStore.AddTag(ref, imageID, true); err != nil {
+			return false, err
+		}
 
+		// Best effort store the digest, this could fail if the digest was
+		// already saved from a different tag.
+		digest, err := reference.WithDigest(ref, manifestDigest)
+		if err != nil {
+			logrus.Errorf("Failed to store digest reference %s: %s", manifestDigest, err.Error())
```
Comment (dmcgowan) [inline distribution/pull_v2.go]: This message should go with the call to `AddDigest` to log an error if one occurred. An error creating the reference would indicate that a reference or digest is bad, which _shouldn't_ be possible here, just returning the error should be fine.

File: distribution/push_v2.go [hunk 4]
```diff
@@ -199,6 +199,15 @@ func (p *v2Pusher) pushV2Tag(ctx context.Context, ref reference.NamedTagged, ima
 
 	manifestDigest := digest.FromBytes(canonicalManifest)
 	progress.Messagef(p.config.ProgressOutput, "", "%s: digest: %s size: %d", ref.Tag(), manifestDigest, len(canonicalManifest))
+
+	// Best effort store the digest
+	digest, err := reference.WithDigest(ref, manifestDigest)
+	if err != nil {
+		logrus.Errorf("Failed to store digest reference %s: %s", manifestDigest, err.Error())
```
Comment (dmcgowan) [inline distribution/push_v2.go]: Same as previous comment, the `reference.WithDigest` call is only making the digest reference and should not fail. The error to be logged is on `AddDigest`

**Comment by dnephin** [comment]
Thanks, makes sense. I've fixed the error handling.

**Comment by dmcgowan** [comment]
I noticed the case where the tag already exists but there is no digest will not cause the digest to be added. This could probably be fixed by always adding the digest reference before the call to `AddTag`.

**Comment by dnephin** [comment]
makes sense, I made that change

**Comment by dnephin** [comment]
Replaced by #23996

**Pull request closed without merging**